### PR TITLE
Generate: visit non-llm `prepare_inputs_for_generation`

### LINKED ISSUE #32685: tracker: move `prepare_inputs_for_generation` into the generation mixin 🧹
🧹 This is a tracker regarding the move of `prepare_inputs_for_generation` into the generation mixin 🧹 

# Why?

1. `prepare_inputs_for_generation` is not part of the core modeling, but rather a utility for `generate`
2. it should greatly reduce the need to touch modeling code, on `generate` changes. Fewer modeling changes -> improved model stability
3. greatly reduced number of lines of code 🙏 

# Tracker

Kinda ordered list of tasks:

- [x] 1. Fix related slow tests before we start — all `llama`, `generate`, and `cache_utils` [except sink cache, broken atm] slow tests should be passing to ensure we don’t break anything (#33138)
- [x] 2. `PreTrainedModel` doesn't inherit from `GenerationMixin`, so that `can_generate()` becomes independent of `prepare_inputs_for_generation` being overwritten or not (#33203)
- [x] 3. Move llama’s `prepare_inputs_for_generation` to the generation mixin. This implies moving one function that prepares the 4D mask too (the one that is called there) (#33677)
- [x] 4. Add tests for the generalist `prepare_inputs_for_generation` — currently we don’t test it directly, and we should (decoder-only llms: https://github.com/huggingface/transformers/pull/33870, encoder-decoder llms: https://github.com/huggingface/transformers/pull/34048)
- [x] 5. Address the case of `synced_gpus` in `generate`: when `synced_gpus` and `cache_positions` is out of bounds, take the latest available `input_ids` for dummy computations (https://github.com/huggingface/transformers/pull/34095)
- [x] 6. Delete `prepare_inputs_for_generation` from as many models as possible. There may be merge conflicts here, due to the 4D mask function. Try to iron out as many trivial cases as possible (decoder-only llms: https://github.com/huggingface/transformers/pull/33870, encoder-decoder llms: https://github.com/huggingface/transformers/pull/34048)
- [x] 7. Change `prepare_inputs_for_generation` to forward `**kwargs` from its input to its output. With minimal changes, this should enable most VLMs to use the shared function -- they forward `pixel_values` from the input to the output (support for **kwargs https://github.com/huggingface/transformers/pull/33870)
- [x] 8. By this point most cases of `prepare_inputs_for_generation` should have been removed 🤗  We would need to check the others individually, there may be further simplification patterns available!

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Closes #32685 🙌 

This PR does a final pass over the remaining `prepare_inputs_for_generation`:
- makes a few adjustments to the general function to handle trivial corner cases
- removes the cases where the general function is equivalent
- adds a comment on the functions that can't be removed, so we can quickly a) remember that there is a general function b) why the general function doesn't work on the model

👉 After this PR, let's aim at overwriting `prepare_inputs_for_generate` as few times as possible, so we can quickly roll out model-agnostic upgrades 🏎️ and minimize bugs 🐛

### CODE REVIEW

**Comment by gante** [comment]
@zucchini-nlp / @ylacombe : i've tagged you both so that yoach can double-check audio models, and raushan the others 🤗

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -806,6 +806,7 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
         device: torch.device,
         cache_position: torch.Tensor,
         batch_size: int,
+        **kwargs,
```
Comment (gante) [inline src/transformers/models/bloom/modeling_bloom.py:809]: Some models have extra kwargs. With `**kwargs` we can make the generalization :)

**Review by ArthurZucker — APPROVED** [review]
🧼 super clean!

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -444,10 +449,12 @@ def prepare_inputs_for_generation(
                     attention_mask,
                     sequence_length=sequence_length,
                     target_length=past_key_values.get_max_cache_shape(),
-                    dtype=self.get_output_embeddings().weight.dtype,
+                    dtype=self.dtype,
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py:452]: for my own understanding, is this needed for multi-gpu setting?

File: src/transformers/models/idefics2/modeling_idefics2.py [hunk 3]
```diff
@@ -1665,6 +1665,9 @@ def prepare_inputs_for_generation(
         num_logits_to_keep=None,
         **kwargs,
     ):
+        # Overwritten -- there are mutually exclusive inputs (if the logic to make `image_hidden_states` take
+        # precedence is moved to the model, we can remove this fn)
+
```
Comment (zucchini-nlp) [inline src/transformers/models/idefics2/modeling_idefics2.py:1670]: I think most VLMs do smth similar when we pass pixel values only in pre-fill stage. I am thinking that for Idefics it can also be a check on `if cache_position[0] == 0` as we don't support multi-turn dialogues. So I am think we can find a way to generalize for VLMs in a subsequent PR :)

About moving the logic to modeling, I think we want to discourage anyone to pass both

**Review by zucchini-nlp — APPROVED** [review]
Great clean up!

**Review by ylacombe — APPROVED** [review]
LGTM, thanks for this @gante! 
Most of the audio models' `prepare_inputs_for_generation` were untouched, but for the ones that were, it looks like it'll work.

Special thanks for Moshi, the new  `prepare_inputs_for_generation` covers every small edge case (no base model, no `get_output_embeddings`, doing post-processing). It's much cleaner to read now.

Are you planning to run slow tests for every models btw ?

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation/utils.py:452]: This change is needed because `moshi` doesn't have `get_output_embeddings()`, and creating `get_output_embeddings()` there would be a bit ambiguous.

`self.dtype` works just as fine and is more versatile :)

**Comment by gante** [comment]
> Are you planning to run slow tests for every models btw ?

Yes :) Now that's approved, I will be running slow tests before merging

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:452]: oh dtype, not device 😄

**Comment by gante** [comment]
Ran slow tests on all models with significant changes, no regressions -- merging :)
